src/midi/midi_soundcanvas.cpp

```diff
@@ -0,0 +1,729 @@
+// the mixer until the requested numbers of audio frames is met.
+void MidiDeviceSoundCanvas::MixerCallback(const int requested_audio_frames)
+{
+	assert(mixer_channel);
+
+	// Report buffer underruns
+	constexpr auto warning_percent = 5.0f;
+
+	if (const auto percent_full = audio_frame_fifo.GetPercentFull();
+	    percent_full < warning_percent) {
+		static auto iteration = 0;
+		if (iteration++ % 100 == 0) {
+			LOG_WARNING("SOUNDCANVAS: Audio buffer underrun");
+		}
+		had_underruns = true;
+	}
+
+	static std::vector<AudioFrame> audio_frames = {};
+
+	const auto has_dequeued = audio_frame_fifo.BulkDequeue(audio_frames,
+	                                                       requested_audio_frames);
+
+	if (has_dequeued) {
+		mixer_channel->AddSamples_sfloat(requested_audio_frames,
+		                                 &audio_frames[0][0]);
+
+		last_rendered_ms = PIC_FullIndex();
+	} else {
+		assert(!audio_frame_fifo.IsRunning());
+		mixer_channel->AddSilence();
+	}
+}
+
+void MidiDeviceSoundCanvas::RenderAudioFramesToFifo(const int num_audio_frames)
+{
+	assert(num_audio_frames > 0);
+
+	static std::vector<float> left  = {};
+	static std::vector<float> right = {};
+
+	// Maybe expand the vectors
+	if (check_cast<int>(left.size()) < num_audio_frames) {
+		left.resize(num_audio_frames);
+		right.resize(num_audio_frames);
+	}
+
+	float* audio_out[] = {left.data(), right.data()};
+
+	clap.plugin->Process(audio_out, num_audio_frames, clap.event_list);
+	clap.event_list.Clear();
+
+	for (auto i = 0; i < num_audio_frames; ++i) {
+		audio_frame_fifo.Enqueue({left[i], right[i]});
+	}
+}
+
+// The next MIDI work task is processed, which includes rendering audio frames
+// prior to sending channel and sysex messages to the plugin
+void MidiDeviceSoundCanvas::ProcessWorkFromFifo()
+{
+	const auto work = work_fifo.Dequeue();
+	if (!work) {
+		return;
+	}
+
+#if 0
+	// To log inter-cycle rendering
+	if (work->num_pending_audio_frames > 0) {
+		LOG_MSG("SOUNDCANVAS: %2u audio frames prior to %s message, followed by "
+		        "%2lu more messages. Have %4lu audio frames queued",
+		        work->num_pending_audio_frames,
+		        work->message_type == MessageType::Channel ? "channel" : "sysex",
+		        work_fifo.Size(),
+		        audio_frame_fifo.Size());
+	}
+#endif
+
+	if (work->num_pending_audio_frames > 0) {
+		RenderAudioFramesToFifo(work->num_pending_audio_frames);
+	}
+
+	if (work->message_type == MessageType::Channel) {
+		assert(work->message.size() >= MaxMidiMessageLen);
+		clap.event_list.AddMidiEvent(work->message, 0);
+
+	} else {
+		assert(work->message_type == MessageType::SysEx);
+		clap.event_list.AddMidiSysExEvent(work->message, 0);
+	}
+}
+
+// Keep the FIFO populated with freshly rendered buffers
+void MidiDeviceSoundCanvas::Render()
+{
+	while (work_fifo.IsRunning()) {
+		work_fifo.IsEmpty() ? RenderAudioFramesToFifo()
+		                    : ProcessWorkFromFifo();
+	}
```

Comment:
`RenderAudioFramesToFifo()` is using a default argument of 1 frame here.  I don't know if Nuked or your CLAP wrapper batches these calls up but it seems rather inefficient to render 1 frame a time.

If the emulation really works on a per-frame basis, maybe it's fine but most audio code I've seen wants to work on at least a couple hundred frames at a time.

The DOSBox mixer callback pulls ~10ms worth of audio at a time depending on blocksize.  Would it be better to base this on blocksize?

Reply:
The emulation is sample based, so buffering here would make zero difference. The cost of a function call is dwarfed by the cost of rendering a single frame. Also, audio code is quite "slow rate", I do not worry about function call costs at "audio rates" at all.

In detail:

- The CLAP host requests a fixed number of frames to be rendered from the plugin.
- Then in essence I need to keep moving the emulation ticking along one MCU clock cycle at a time with `MCU_Step()` until it spits out the number of frames we want (the number of MCU steps needed is hard to determine in advance... I tried and failed spectacularly, I don't even think now it's possible as it "jitters", so any estimates will slowly get out of sync over time). See here https://github.com/johnnovak/Nuked-SC55-CLAP/blob/main/src/nuked_sc55.cpp#L437

So it's actually "worse" than what you thought; we need to run the MCU emulation for several (variable number per sample frame) of clock cycles until it publishes a single sample frame into the out buffer (more cycles per sample frame are needed for the mk2; that's one of the reasons why emulating that model is more demanding).

Reply:
Yeah, the cost of a function call is almost negligible.  I was more wondering about the SC55 internals.  That's a lot of work that goes into one audio frame then!

I just benchmarked it out of curiosity.  It takes around 13 microseconds average for the 1 frame calls and it scales up almost perfectly linearly with the larger calls.  So OK, probably not going to make much of a difference how many frames we request at one time.

Reply:
> Yeah, the cost of a function call is almost negligible. I was more wondering about the SC55 internals. That's a lot of work that goes into one audio frame then!

Yeah because it's low-level emulation, almost "simulation". Most of the various chips from the hardware are fully emulated, so they run the real firmware on that custom CPU the Sound Canvas has, and so on.